File: dmoj/ccc23s5/ccc23s5_bitset.py
n = int(input())
PW = 3**18

MN = int(1e9)
bits = bytearray(MN>>3)
to_clear = []

def check(x):
    global bits
    for a in to_clear:
        bits[a>>3] ^= 1 << (a&7)
    to_clear.clear()

    while True:
        d = x//n
        x -= d*n
        if d == 1 and x: return False
        if bits[x>>3] & (1<<(x&7)): return True
        bits[x>>3] |= 1<<(x&7)
        to_clear.append(x)
        x *= 3

def solve(l, r):
    dif = r-l
    if dif > 1:
        solve(l, l + dif//3)
        solve(r - dif//3, r)
    else:
        a = (n*l + PW-1)//PW
        b = n*r//PW + 1
        for i in range(a, b):
            if check(i): print(i)

solve(0, PW)

File: dmoj/ccc23s5/ccc23s5_bitset.cpp
#include <bits/stdc++.h>
using namespace std;
typedef unsigned int uint;
typedef long long ll;

const int MN = 1e9;
const int PW = 1162261467;
int n;
bitset<MN> bits;
vector<int> vec;

bool check(uint x){
    for(int a : vec) bits[a] = false;
    vec.clear();
    while(true){
        uint d = x/n;
        x -= n*d;
        if(d == 1 && x) return false;
        if(bits[x]) return true;
        bits[x] = true;
        vec.push_back(x);
        x *= 3;
    }
}

void solve(int l, int r){
    int len = r-l;
    if(len > 1) solve(l, l + len/3), solve(r - len/3, r);
    else {
        int a = ((ll) l*n + PW-1)/PW;
        int b = ((ll) r*n + PW)/PW;
        for(int i = a; i < b; i++) if(check(i)) cout << i << '\n';
    }
}

int main(){
    cin.sync_with_stdio(0); cin.tie(0);
    cin >> n;
    solve(0, PW);
}
